code/core_save.c

```diff
@@ -23,67 +23,70 @@ void save_galaxies(int filenr, int tree)
   int OutputGalCount[MAXSNAPS], *OutputGalOrder;
 
   OutputGalOrder = (int*)malloc( NumGals*sizeof(int) );
-  assert( OutputGalOrder );
+  if(OutputGalOrder == NULL) {
+    fprintf(stderr,"Error: Could not allocate memory for %d int elements in array `OutputGalOrder`\n", NumGals);
+    ABORT(10);
+  }
 
   // reset the output galaxy count and order
   for(i = 0; i < MAXSNAPS; i++)
     OutputGalCount[i] = 0;
   for(i = 0; i < NumGals; i++)
     OutputGalOrder[i] = -1;
-
+  
   // first update mergeIntoID to point to the correct galaxy in the output
   for(n = 0; n < NOUT; n++)
-  {
-    for(i = 0; i < NumGals; i++)
     {
-      if(HaloGal[i].SnapNum == ListOutputSnaps[n])
-      {
-        OutputGalOrder[i] = OutputGalCount[n];
-        OutputGalCount[n]++;
-      }
+      for(i = 0; i < NumGals; i++)
+	{
+	  if(HaloGal[i].SnapNum == ListOutputSnaps[n])
+	    {
+	      OutputGalOrder[i] = OutputGalCount[n];
+	      OutputGalCount[n]++;
+	    }
+	}
     }
-  }
-    
+  
   for(i = 0; i < NumGals; i++)
     if(HaloGal[i].mergeIntoID > -1)
       HaloGal[i].mergeIntoID = OutputGalOrder[HaloGal[i].mergeIntoID];    
   
   // now prepare and write galaxies
   for(n = 0; n < NOUT; n++)
-  {
-    // only open the file if it is not already open.
-    if( !save_fd[n] )
-    {
-      sprintf(buf, "%s/%s_z%1.3f_%d", OutputDir, FileNameGalaxies, ZZ[ListOutputSnaps[n]], filenr);
-
-      if(!(save_fd[n] = fopen(buf, "r+")))
-      {
-				printf("can't open file `%s'\n", buf);
-				ABORT(0);
-      }
-
-			// write out placeholders for the header data.
-			size_t size = (Ntrees + 2)*sizeof(int);
-			int* tmp_buf = (int*)malloc( size );
-			memset( tmp_buf, 0, size );
-			fwrite( tmp_buf, sizeof(int), Ntrees + 2, save_fd[n] );
-			free( tmp_buf );
-		}
-
-    for(i = 0; i < NumGals; i++)
     {
-      if(HaloGal[i].SnapNum == ListOutputSnaps[n])
-      {        
-        prepare_galaxy_for_output(filenr, tree, &HaloGal[i], &galaxy_output);
-        myfwrite(&galaxy_output, sizeof(struct GALAXY_OUTPUT), 1, save_fd[n]);
-
-        TotGalaxies[n]++;
-        TreeNgals[n][tree]++;
-      }
+      // only open the file if it is not already open.
+      if( !save_fd[n] )
+	{
+	  sprintf(buf, "%s/%s_z%1.3f_%d", OutputDir, FileNameGalaxies, ZZ[ListOutputSnaps[n]], filenr);
+	  
+	  if(!(save_fd[n] = fopen(buf, "r+")))
+	    {
+	      printf("can't open file `%s'\n", buf);
+	      ABORT(0);
+	    }
+	  
+	  // write out placeholders for the header data.
+	  size_t size = (Ntrees + 2)*sizeof(int);
+	  int* tmp_buf = (int*)malloc( size );
+	  memset( tmp_buf, 0, size );
+	  fwrite( tmp_buf, sizeof(int), Ntrees + 2, save_fd[n] );
```

Comment:
Please check the output of `malloc` and `fwrite`;  unlike `free`, `malloc` and `fwrite` are not guaranteed. 

Well, `malloc` can not fail for `Too Small to Fail` memory requests, with the limit determined by by `PAGE_ALLOC_COSTLY_ORDER`. A few times the typical memory page size (4 KB) is should always succeed, otherwise the OS goes around killing random processes to reclaim memory (`Out of Memory Killer`, I am serious - that's the name). See [here](https://lwn.net/Articles/627419/) for details.